src/Messages/OWSIdentityManager.m

```diff
@@ -361,6 +398,286 @@ - (void)createIdentityChangeInfoMessageForRecipientId:(NSString *)recipientId
+                [self sendSyncVerificationStateMessage:message];
+            }
+        }
+    });
+}
+
+- (void)syncAllVerificationStates
+{
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        @synchronized(self)
+        {
+            OWSVerificationStateSyncMessage *message =
+            [OWSVerificationStateSyncMessage new];
+            [OWSRecipientIdentity enumerateCollectionObjectsUsingBlock:^(OWSRecipientIdentity *recipientIdentity, BOOL *stop) {
+                OWSAssert(recipientIdentity);
+                OWSAssert(recipientIdentity.recipientId.length > 0);
+                OWSAssert(recipientIdentity.identityKey.length > 0);
+
+                if (recipientIdentity.recipientId.length < 1 || recipientIdentity.identityKey.length < 1) {
+                    OWSFail(@"Invalid recipient identity for recipientId: %@", recipientIdentity.recipientId);
+                    return;
+                }
+                [message addVerificationState:recipientIdentity.verificationState
+                                  identityKey:recipientIdentity.identityKey
+                                  recipientId:recipientIdentity.recipientId];
+            }];
+            if (message.recipientIds.count > 0) {
+                [self sendSyncVerificationStateMessage:message];
+            }
+        }
+    });
+}
+
+- (void)sendSyncVerificationStateMessage:(OWSVerificationStateSyncMessage *)message
+{
+    OWSAssert(message);
+    OWSAssert(message.recipientIds.count > 0);
+
+    [self.messageSender sendMessage:message
+        success:^{
+            DDLogInfo(@"%@ Successfully sent verification state sync message", self.tag);
+
+            // Record that this verification state was successfully synced.
+            [self clearSyncMessageForRecipientIds:message.recipientIds];
+        }
+        failure:^(NSError *error) {
+            DDLogError(@"%@ Failed to send verification state sync message with error: %@", self.tag, error);
+        }];
+}
+
+- (void)clearSyncMessageForRecipientId:(NSString *)recipientId
+{
+    OWSAssert(recipientId.length > 0);
+
+    [self clearSyncMessageForRecipientIds:@[recipientId]];
+}
+
+- (void)clearSyncMessageForRecipientIds:(NSArray<NSString *> *)recipientIds
+{
+    OWSAssert(recipientIds.count > 0);
+
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        @synchronized(self)
+        {
+            for (NSString *recipientId in recipientIds) {
+                [self.storageManager removeObjectForKey:recipientId
+                                           inCollection:OWSIdentityManager_QueuedVerificationStateSyncMessages];
+            }
+        }
+    });
+}
+
+- (void)processIncomingSyncMessage:(NSArray<OWSSignalServiceProtosSyncMessageVerification *> *)verifications
+{
+    for (OWSSignalServiceProtosSyncMessageVerification *verification in verifications) {
+        NSString *recipientId = [verification destination];
```

Comment:
nit: I think regular property getters are generally more readable. Sometimes I use the message-send style when I'm trying to convey something as more expensive than a simple "get" but... that's probably just a silly thing I do.